### DISCUSSION TITLE
V8ScriptEngine usage pattern

### DISCUSSION THREAD

**Discussion by IkeOTL** [ClearFoundry/ClearScript#355] [Q&A] [answered]
I have a few questions about using V8ScriptEngine properly. I have an environment that will be polling a queue for script names, and then will download/cache/compile those scripts by name.

My questions are: 
- Is it safe to use a compiled V8Script with another instance of V8ScriptEngine.
- Is it safe to reuse an instance of V8ScriptEngine with multiple compiled scripts with the same function names?
- Is it best to simply create a new instance of V8ScriptEngine every time I need to execute a V8Script and invoke a function in the script?

Thanks in advance,
Ike

**Comment by ClearScriptLib** [comment, marked as answer]
Hello Ike,

> Is it safe to use a compiled V8Script with another instance of V8ScriptEngine.

That's safe only if the script engines share a [V8 runtime](https://clearscript.clearfoundry.net/Reference/html/T_Microsoft_ClearScript_V8_V8Runtime.htm). Attempting to execute a compiled script in a foreign runtime should generate an exception, although there's currently a bug that causes a crash in this situation.

> Is it safe to reuse an instance of V8ScriptEngine with multiple compiled scripts with the same function names?

If you mean that the scripts create global functions with the same names, then yes, that's safe.

> Is it best to simply create a new instance of V8ScriptEngine every time I need to execute a V8Script and invoke a function in the script?

Creating a new engine instance is the only way to guarantee a pristine JavaScript environment. However, instantiating `V8ScriptEngine` is relatively expensive, especially if the instances don't share a V8 runtime. Therefore, the best strategy depends on your performance requirements and usage patterns. For example, if you can trust your scripts not to corrupt the environment, then reusing engine instances should be perfectly safe.

Please don't hesitate to send additional questions our way.

Thank you!
